### ISSUE TITLE
Moq as TestCaseSource dosen't work

### ISSUE THREAD

**Issue by meriturva** [nunit/nunit#841]
Hi to all,
we use TestCaseSource with Mock instances and we found that the test is not called.
We use the NUnit 3 beta 4 version with Visual Studio 2015 Community Edition.
This is just a sample code to reproduce the problem:

```C#
    public class Value
    {
        public double RowVale { get; set; }
    }

    [TestFixture]
    public class TestTest
    {
        private static Value[] Values = new Test.Value[2];
        
        static TestTest()
        {
            Values[0] = new Value()
            {
                RowVale = 10
            };

            var mockValue = new Mock<Value>();
            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(Value value)
        {
           
        }

    }
```

**Comment by rprouse** [comment]
Thanks for the report. I have set the milestone to the next release so that we can confirm and triage the issue.

**Comment by CharliePoole** [comment]
The test cases are created at the time your tests are loaded, using a mock created at that time. Then, when the tests are executed, a new object is created and Values array is re-initialized. However, that initialization cannot impact the cases that are run because they are already defined.

It could be that Moq introduces a new variable into the problem but it really comes down to a standard issue/feature of NUnit: that test case initialization is static and can't be affected by anything that takes place during the execution of a test run. We already have plans to add dynamic test cases in a future release.

I'd be inclined to close this as wontfix or notabug.

**Comment by meriturva** [comment]
The main problem here is that the Mock is created by Static Constructor and the array is already declared with the correct type of the method.
I think the main problem is how NUnit looking for the right test method to call in case of TestCaseSource.
If you still, in future, permit static and dynamic TestCaseSource i guess it will be important to fix that case!
If you just move all TestCaseSource to dynamic initialization the story changes......but just check that it works with Mock objects.
Thanks.

**Comment by constructor-igor** [comment]
please, could you help to understand
why next sample
```c#
using System;
using NUnit.Framework;

namespace NUnit_v3_samples
{
    public class Value
    {
        public double RowVale { get; set; }
    }
    [TestFixture]
    public class NUnitIssue841
    {
        private static Value[] Values = new Value[2];

        static NUnitIssue841()
        {
            Values[0] = new Value
            {
                RowVale = 10
            };
            Values[1] = new Value
            {
                RowVale = 20
            };

//            var mockValue = new Mock<Value>();
//            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(Value value)
        {
            Console.WriteLine("RowVale: {0}", value.RowVale);
        }
    }
}
```
shows single test only:
![image](https://cloud.githubusercontent.com/assets/1849690/10304033/510f944c-6c20-11e5-8520-982250da82d2.png)

Also, the issue has status "status:confirm". Can I help here?

**Comment by CharliePoole** [comment]
My guess is that there are two results shown in the adapter.

The "status:confirm" means we need to verify that the problem actually exists and in this case try to make it more precise. In this case, that would include:

* Is it really an NUnit issue? It's reported under the nunit project but the description talks about using the adapter - no version of that specified. The first step would be to run it under the NUnit console runner to see what happens.

* Is it really a Moq issue. If we can duplicate it without use of a mock, then that's one thing. If it only happens under Moq, then it's something else.

Any light you could shed on it will help. We'll still have to decide whether we can fix it for the RC or not, depending on how much work is involved.

**Comment by constructor-igor** [comment]
NUnit console:

```c#
using System;
using Moq;
using NUnit.Framework;

namespace NUnit_v3_samples
{
    public class Value
    {
        public double RowVale { get; set; }
    }
    [TestFixture]
    public class NUnitIssue841
    {
        private static Value[] Values = new Value[2];

        static NUnitIssue841()
        {
            Values[0] = new Value
            {
                RowVale = 10
            };
//            Values[1] = new Value
//            {
//                RowVale = 20
//            };

            var mockValue = new Mock<Value>();
            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(Value value)
        {
            Console.WriteLine("RowVale: {0}", value.RowVale);
        }
    }
}
```

result:
```xml
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="40" result="Passed" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.125457">
  <environment nunit-version="3.0.5715.30860" clr-version="4.0.30319.34209" os-version="Microsoft Windows NT 6.1.7601 Service Pack 1" platform="Win32NT" cwd="D:\My\@gh\TechSugar\TestingLab\NUnitSamples\NUnit_v3_samples" machine-name="IL-IGOR-Z-LW7" user="igor-z" user-domain="NOVA" culture="en-US" uiculture="en-US" os-architecture="x64" />
  <test-suite type="Assembly" id="0-1079" name="NUnit_v3_samples.dll" fullname="D:\My\@gh\TechSugar\TestingLab\NUnitSamples\NUnit_v3_samples\bin\debug\NUnit_v3_samples.dll" runstate="Runnable" testcasecount="40" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.057166" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="_PID" value="9980" />
      <property name="_APPDOMAIN" value="test-domain-" />
    </properties>
    <test-suite type="SetUpFixture" id="0-1046" name="NUnit_v3_samples" fullname="NUnit_v3_samples.SetUpFixture" classname="NUnit_v3_samples.SetUpFixture" runstate="Runnable" testcasecount="40" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.037159" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
      <output>SetUpFixture.OneTimeSetUp

SetUpFixture.OneTimeSetUp
SetUpFixture.OneTimeTearDown

</output>
      <test-suite type="TestFixture" id="0-1039" name="NUnitIssue841" fullname="NUnit_v3_samples.NUnitIssue841" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" testcasecount="2" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.028931" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
        <test-suite type="ParameterizedMethod" id="0-1042" name="TestTestCaseSourceWithMoq" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq" runstate="Runnable" testcasecount="2" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.028484" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
          <test-case id="0-1040" name="TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" methodname="TestTestCaseSourceWithMoq" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" seed="1960734727" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.008601" asserts="0">
            <output>RowVale: 10
</output>
          </test-case>
          <test-case id="0-1041" name="TestTestCaseSourceWithMoq(Mock&lt;NUnit_v3_samples.Value:8cde&gt;.Object)" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq(Mock&lt;NUnit_v3_samples.Value:8cde&gt;.Object)" methodname="TestTestCaseSourceWithMoq" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" seed="931638115" result="Passed" start-time="2015-10-06 10:22:59Z" end-time="2015-10-06 10:22:59Z" duration="0.000061" asserts="0">
            <output>RowVale: 0
</output>
          </test-case>
        </test-suite>
      </test-suite>
    </test-suite>
  </test-suite>
</test-run>
```

**Comment by constructor-igor** [comment]
test without mock:
```c#
using System;
using Moq;
using NUnit.Framework;

namespace NUnit_v3_samples
{
    public class Value
    {
        public double RowVale { get; set; }
    }
    [TestFixture]
    public class NUnitIssue841
    {
        private static Value[] Values = new Value[2];

        static NUnitIssue841()
        {
            Values[0] = new Value
            {
                RowVale = 10
            };
            Values[1] = new Value
            {
                RowVale = 20
            };

//            var mockValue = new Mock<Value>();
//            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(Value value)
        {
            Console.WriteLine("RowVale: {0}", value.RowVale);
        }
    }
}
```

Result:
```xml
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="40" result="Passed" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.108611">
  <environment nunit-version="3.0.5715.30860" clr-version="4.0.30319.34209" os-version="Microsoft Windows NT 6.1.7601 Service Pack 1" platform="Win32NT" cwd="D:\My\@gh\TechSugar\TestingLab\NUnitSamples\NUnit_v3_samples" machine-name="IL-IGOR-Z-LW7" user="igor-z" user-domain="NOVA" culture="en-US" uiculture="en-US" os-architecture="x64" />
  <test-suite type="Assembly" id="0-1079" name="NUnit_v3_samples.dll" fullname="D:\My\@gh\TechSugar\TestingLab\NUnitSamples\NUnit_v3_samples\bin\debug\NUnit_v3_samples.dll" runstate="Runnable" testcasecount="40" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.045669" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="_PID" value="11532" />
      <property name="_APPDOMAIN" value="test-domain-" />
    </properties>
    <test-suite type="SetUpFixture" id="0-1046" name="NUnit_v3_samples" fullname="NUnit_v3_samples.SetUpFixture" classname="NUnit_v3_samples.SetUpFixture" runstate="Runnable" testcasecount="40" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.027310" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
      <output>SetUpFixture.OneTimeSetUp

SetUpFixture.OneTimeSetUp
SetUpFixture.OneTimeTearDown

</output>
      <test-suite type="TestFixture" id="0-1039" name="NUnitIssue841" fullname="NUnit_v3_samples.NUnitIssue841" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" testcasecount="2" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.019234" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
        <test-suite type="ParameterizedMethod" id="0-1042" name="TestTestCaseSourceWithMoq" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq" runstate="Runnable" testcasecount="2" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.018770" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
          <test-case id="0-1040" name="TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" methodname="TestTestCaseSourceWithMoq" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" seed="1960734727" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.008267" asserts="0">
            <output>RowVale: 10
</output>
          </test-case>
          <test-case id="0-1041" name="TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" fullname="NUnit_v3_samples.NUnitIssue841.TestTestCaseSourceWithMoq(NUnit_v3_samples.Value)" methodname="TestTestCaseSourceWithMoq" classname="NUnit_v3_samples.NUnitIssue841" runstate="Runnable" seed="931638115" result="Passed" start-time="2015-10-06 10:26:52Z" end-time="2015-10-06 10:26:52Z" duration="0.000059" asserts="0">
            <output>RowVale: 20
</output>
          </test-case>
        </test-suite>
      </test-suite>
    </test-suite>
  </test-suite>
</test-run>
```

**Comment by constructor-igor** [comment]
Short report:
NUnit v3beta4 console executes 2 tests without unexpected issues
NUnit v3 VS2013 Adapter execute one test only.

**Comment by CharliePoole** [comment]
So, in the adapter, does it fail without regard to use of mock? Does it not show one test with two results?

One test with two results might be expected if the test names created were the same for both of them. The results would show at the bottom of the Test window when the test is selected.

**Comment by constructor-igor** [comment]
test with mock
```c#
using System;
using Moq;
using NUnit.Framework;

namespace NUnit_v3_samples
{
    public interface IValue
    {
        double RowVale { get; set; }
    }
    public class Value: IValue
    {
        public double RowVale { get; set; }
    }
    [TestFixture]
    public class NUnitIssue841
    {
        private static IValue[] Values = new IValue[2];

        static NUnitIssue841()
        {
            Values[0] = new Value
            {
                RowVale = 10
            };
//            Values[1] = new Value
//            {
//                RowVale = 20
//            };

            var mockValue = new Mock<IValue>();
            mockValue.Setup(foo => foo.RowVale).Returns(30.0);
            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(IValue value)
        {
            Console.WriteLine("RowVale: {0}", value.RowVale);
        }
    }
}
```

Adapter shows 2 tests (first - executed, seconds - unknown status)
![image](https://cloud.githubusercontent.com/assets/1849690/10306962/68739206-6c33-11e5-926f-3b6eafb47649.png)

**Comment by constructor-igor** [comment]
Next test (without mock)
```c#
using System;
using Moq;
using NUnit.Framework;

namespace NUnit_v3_samples
{
    public interface IValue
    {
        double RowVale { get; set; }
    }
    public class Value: IValue
    {
        public double RowVale { get; set; }
    }
    [TestFixture]
    public class NUnitIssue841
    {
        private static IValue[] Values = new IValue[2];

        static NUnitIssue841()
        {
            Values[0] = new Value
            {
                RowVale = 10
            };
            Values[1] = new Value
            {
                RowVale = 20
            };

//            var mockValue = new Mock<IValue>();
//            mockValue.Setup(foo => foo.RowVale).Returns(30.0);
//            Values[1] = mockValue.Object;
        }

        [Test, TestCaseSource("Values")]
        public void TestTestCaseSourceWithMoq(IValue value)
        {
            Console.WriteLine("RowVale: {0}", value.RowVale);
        }
    }

    [TestFixture]
    public class NUnitIssue841_Moq
    {
        [Test]
        public void TestMoq()
        {
            var mockValue = new Mock<IValue>();
            mockValue.Setup(foo => foo.RowVale).Returns(30.0);
            Assert.That(mockValue.Object.RowVale, Is.EqualTo(30.0));
        }
    }
}
```
shows one test with 2 results:
![image](https://cloud.githubusercontent.com/assets/1849690/10307004/b2c63304-6c33-11e5-9377-f3dbb1bf9d4e.png)

**Comment by x789** [comment]
I faced the same behaviour today using the 2.0 version of the NUnit test adapter for VS2013. But even if the adapter would support a mock created with Moq the name of the test would change each time it is executed.

To work around this behaviour I use NUnit's TestCaseData class to provide the test-data and the name of the test-case. This name does not change between initialization and execution so the NUnit test adapter executs my tests flawlessly. However it has the draw-back that the array used as TestCaseSource can be used only for one test-method (to avoid name collisions).

My code is similar to the following:

    [TestCaseFixture]
    public class MyTest
    {
      private static TestCaseData[] testData = new TestCaseData[] { 
                                                                    new TestCaseData(null).SetName("Test01(null)"),
																    new TestCaseData(new Mock<IMyInterface>().Object).SetName("Test01(with mock)") };
    
      [Test, TestCaseSource("testData")]
      public void Test01(IMyInterface param)
      {
        // Test logic
      }
	}

**Comment by CharliePoole** [comment]
I think you have scoped out the problem here:

1. Static initialization is not causing it.

2. Since the mock objects are created twice - once in discovery and again when running the test - they have different names! We already know we have this problem with random arguments. This is best resolved with what we are calling "dynamic test cases", that is, test cases that are not generated statically, but only at run time. If we generate a test only once, then we won't have this name conflict. There are several issues already addressing this, but it will not be in the 3.0 release.

3. When not using the mocks, you can have situations where two tests have the same name. This doesn't bother NUNit because we identify tests by ID, not name. However, Visual Studio uses the name, which leads to the phenomenon of two results for one test. Issue #59 addresses this, among other things. We will need to detect duplicates and add a differentiating suffix.

4. As @x789 pointed out, use of SetName creates duplicates if you use the same data across several test. These duplicates are for different methods, which makes the reporting particularly hard to follow. This problem will be resolved when we implement the new name generation spec, documented at https://github.com/nunit/dev/wiki/Generation-of-Test-Names You will be able to use a name like "{m}(test 01)" causing the actual method name to be used.

I took some time to analyze all this because I'd like to close this issue as notabug or duplicate. Do you guys agree that there are no new problems identified or do you think we need to keep this issue open (or create a new one for the adapter).

**Comment by x789** [comment]
I think this is neither a bug in NUnit nor the test adapter. Its a kind of technical limitation.
Have you already decided which version of NUnit will have the new name generation?

**Comment by CharliePoole** [comment]
It's listed for 3.0. Of course, it could be pushed back, but I've given it a high priority.

I'm going to go ahead and close this as duplicating #59, #68 and #141 (at least the discussion here touched on all those).
